src/search.cpp

```diff
@@ -379,29 +356,17 @@ Search::iterator Search::begin() const {
         }
     }
 
-    enquire.set_query(query);
-
-#if WITH_LEV
-    if (suggestion_mode && !hasNewSuggestionFormat) {
-      size_t value_index = 0;
-      bool has_custom_distance_maker = true;
-      if ( !valuesmap.empty() ) {
-        if ( valuesmap.find("title") != valuesmap.end() ) {
-          value_index = valuesmap["title"];
-        } else {
-          // This should not happen as valuesmap has a title entry, but let's
-          // be tolerent.
-          has_custom_distance_maker = false;
-        }
-      }
-      auto temp_results = enquire.get_mset(0,0);
-      if ( has_custom_distance_maker
-        && temp_results.get_matches_estimated() <= MAX_MATCHES_TO_SORT ) {
-        keyMaker.reset(new LevenshteinDistanceMaker(this->query, value_index));
-        enquire.set_sort_by_key(keyMaker.get(), false);
-      }
+    // In suggestion mode, we are searching over a separate title index.
+    // Default BM25 is not adapted for this case, tuning down wdf factor
+    // (k1) and increasing length normalization factor(b) is necessary.
+    // The document set is first sorted by their relevance score then by 
+    // value so that suggestion results are closer to search string.
+    if (suggestion_mode) {
+      enquire.set_weighting_scheme(Xapian::BM25Weight(0.001,0,1,1,0.5));
+      enquire.set_sort_by_relevance_then_value(valuesmap["title"], false);
```

Comment:
Those values are pretty obscure.
Why do we need to change `k1` and `b` ? and why to this value ?
Do we need to change other factors ?

I suppose that you have read some documentation to find this, it would be nice to have a link to it. Or document the tuning here (or in commit message)

Reply:
@mgautierfr Yes, doc is here https://xapian.org/docs/apidoc/html/classXapian_1_1Weight.html... but not sure this is a good idea to start to put documentation links to external doc when this is trivial to find. Anyway maybe here we can do an exception.

Reply:
You are pointing to the api of `Xapian::Weight` class. This class is implemented by `BM25Weight` but never directly used in our code (and it cannot as it is a abstract class).

But I don't ask for this kind of documentation. I'm asking why changing `k1` to `0.001` and `b` to `1` is good. Why those values and not another ones ?
Why not change `k2`, `k3` or `min_normlen` ? Do we have documentation on the `BM25Weight` algorithm (not the api) ?

Reply:
@mgautierfr [This](https://xapian.org/docs/apidoc/html/classXapian_1_1BM25Weight.html) is the documentation for BM25Weight , I should rather add the reason for the value changes as a comment in the code itself. It will be helpful for future references when we optimize the search further.

Reply:
The reason is somehow already in the comment in the code and in the https://github.com/openzim/libzim/pull/492#issue-568802032

What I would like to know is how do you choose `0.001` and `1`. Do you just try it and the result was good enough ? Does a documentation in xapian say that for short content it is better to tune the algorithm this way ? Do the values are the good one or just better than the default (and we may have better ones) ?

(I'm not saying the values are bad. They are pretty good and improve greatly the results. I just want to understand why those values, at least for my personal curiosity)

Reply:
To be honest it was mostly trial and error after understanding the effect of each parameter. There is no documentation specifically for our use case, so I dropped a couple of queries in the xapian irc. There results are compiled [here](https://etherpad.wikimedia.org/p/better_kiwix_suggestions). The reason I ended up with these parameters specifically are mentioned in my last comment. These values can be tuned to even better with some changes to Xapian::PostList, which I intend to work on later. So yes, a better tuning parameter can exist with some modification to the code.